lib/sentry/context.ex

```diff
@@ -1,78 +1,74 @@
-    This imposes some limitations. The state will only exist within 
-    the current process, and the context will die with the process.
+  Sentry.Context uses the Process Dictionary to store this state.
+  This imposes some limitations. The state will only exist within
+  the current process, and the context will die with the process.
 
-    For example, if you add context inside your controller and an 
-    error happens in a Task the context won't send.
+  For example, if you add context inside your controller and an
+  error happens in a Task the context won't send.
   """
+
   @process_dictionary_key :sentry_context
   @user_key :user
   @tags_key :tags
   @extra_key :extra
   @request_key :request
   @breadcrumbs_key :breadcrumbs
 
-  def get_all do
-    context = get_context()
-    %{
-      user: Map.get(context, @user_key, %{}),
-      tags: Map.get(context, @tags_key, %{}),
-      extra: Map.get(context, @extra_key, %{}),
-      request: Map.get(context, @request_key, %{}),
-      breadcrumbs: Map.get(context, @breadcrumbs_key, []),
-    }
-  end
+  @keys [{:user, @user_key},
+         {:tags, @tags_key},
+         {:extra, @extra_key},
+         {:http, @request_key}]
 
-  def set_extra_context(map) when is_map(map) do
-    get_context()
-    |> set_context(@extra_key, map)
-  end
+  @type data :: keyword() | map()
 
-  def set_user_context(map) when is_map(map) do
-    get_context()
-    |> set_context(@user_key, map)
-  end
+  def get_all do
+    context = get_context()
 
-  def set_tags_context(map) when is_map(map) do
-    get_context()
-    |> set_context(@tags_key, map)
+    for {_, key} <- @keys, into: %{} do
+      {key, Map.get(context, key, %{})}
+    end
+    |> Map.put(@breadcrumbs_key, Map.get(context, :breadcrumbs, []))
   end
 
-  def set_http_context(map) when is_map(map) do
-    get_context()
-    |> set_context(@request_key, map)
+  Enum.each @keys, fn {name, key} ->
+    @spec unquote(:"set_#{name}_context")(data):: any()
+    def unquote(:"set_#{name}_context")(map) do
+      map
+      |> Map.new
+      |> set_context(unquote(key))
+    end
   end
 
   def clear_all do
     Process.delete(@process_dictionary_key)
   end
 
   defp get_context do
-    Process.get(@process_dictionary_key) || %{}
+    Process.get(@process_dictionary_key, %{})
   end
 
-  def add_breadcrumb(map) when is_map(map) do
-    map = Map.put_new(map, "timestamp", Sentry.Util.unix_timestamp())
-    current_context = get_context()
-    breadcrumbs = Map.get(current_context, @breadcrumbs_key, [])
-    new_breadcrumbs = breadcrumbs ++ [map]
-    new_context = Map.put(current_context, @breadcrumbs_key, new_breadcrumbs)
-    Process.put(@process_dictionary_key, new_context)
+  @spec add_breadcrumb(data) :: any()
+  def add_breadcrumb(map) do
+    map = map
+          |> Map.new
+          |> Map.put_new(:timestamp, Sentry.Util.unix_timestamp())
+
+    context = get_context()
+              |> Map.update(@breadcrumbs_key, [map], &(&1 ++ [map]))
```

Comment:
Review if Sentry API requires sorted list, if not then `[map | &1]` should be faster than `&1 ++ [map]`.